src/validator/MatchValidator.cpp

```diff
@@ -23,47 +23,113 @@ AstContext* MatchValidator::getAstContext() {
+                }
+                matchCtx_->clauses.emplace_back(std::move(matchClauseCtx));
 
-    auto retClauseCtx = getContext<ReturnClauseContext>();
-    retClauseCtx->aliases = &matchClauseCtx->aliases;
-    NG_RETURN_IF_ERROR(validateReturn(sentence->ret(), *matchClauseCtx, *retClauseCtx));
+                break;
+            }
+            case ReadingClause::Kind::kUnwind: {
+                auto *unwindClause = static_cast<UnwindClause *>(clauses[i].get());
+                auto unwindClauseCtx = getContext<UnwindClauseContext>();
+                unwindClauseCtx->aliasesUsed = aliasesUsed;
+                NG_RETURN_IF_ERROR(validateUnwind(unwindClause, *unwindClauseCtx));
+
+                if (aliasesUsed) {
+                    NG_RETURN_IF_ERROR(
+                        combineAliases(unwindClauseCtx->aliasesGenerated, *aliasesUsed));
+                }
+                aliasesUsed = &unwindClauseCtx->aliasesGenerated;
+                if (prevYieldColumns) {
+                    NG_RETURN_IF_ERROR(combineYieldColumns(
+                        const_cast<YieldColumns *>(unwindClauseCtx->yieldColumns),
+                        prevYieldColumns));
+                }
+                prevYieldColumns = const_cast<YieldColumns *>(unwindClauseCtx->yieldColumns);
+
+                if (i == clauses.size() - 1) {
+                    retClauseCtx->aliasesUsed = aliasesUsed;
+                    NG_RETURN_IF_ERROR(
+                        validateReturn(sentence->ret(), unwindClauseCtx.get(), *retClauseCtx));
+                }
+                matchCtx_->clauses.emplace_back(std::move(unwindClauseCtx));
 
-    matchCtx_->clauses.emplace_back(std::move(matchClauseCtx));
+                break;
+            }
+            case ReadingClause::Kind::kWith: {
+                auto *withClause = static_cast<WithClause *>(clauses[i].get());
+                auto withClauseCtx = getContext<WithClauseContext>();
+                withClauseCtx->aliasesUsed = aliasesUsed;
+                NG_RETURN_IF_ERROR(validateWith(withClause, *withClauseCtx));
+                if (withClause->where() != nullptr) {
+                    auto whereClauseCtx = getContext<WhereClauseContext>();
+                    whereClauseCtx->aliasesUsed = &withClauseCtx->aliasesGenerated;
+                    NG_RETURN_IF_ERROR(
+                        validateFilter(withClause->where()->filter(), *whereClauseCtx));
+                    withClauseCtx->where = std::move(whereClauseCtx);
+                }
+
+                aliasesUsed = &withClauseCtx->aliasesGenerated;
+                prevYieldColumns = const_cast<YieldColumns *>(withClauseCtx->yieldColumns);
+
+                if (i == clauses.size() - 1) {
+                    retClauseCtx->aliasesUsed = aliasesUsed;
+                    NG_RETURN_IF_ERROR(
+                        validateReturn(sentence->ret(), withClauseCtx.get(), *retClauseCtx));
+                }
```

Comment:
If all cases have to vailidate return sentence, why not pull this check outside the switch block?

Reply:
```
NG_RETURN_IF_ERROR(
                        validateReturn(sentence->ret(), withClauseCtx.get(), *retClauseCtx));
matchCtx_->clauses.emplace_back(std::move(withClauseCtx));
```
Because return clause uses the preceding clause to call `validateReturn`, so if pull validateREturn outside the switch block, I can't std::move(the_preceding_clause).